src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,348 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+'use strict';
+
+import * as fs from 'fs';
+import * as path from 'path';
+import { EventEmitter } from 'events';
+import { Deferred } from 'ts-deferred';
+import * as component from '../../../common/component';
+import { getExperimentId } from '../../../common/experimentStartupInfo';
+import { getLogger, Logger } from '../../../common/log';
+import { EnvironmentInformation, EnvironmentService } from '../environment';
+import { NNIError, NNIErrorNames } from '../../../common/errors';
+import { ObservableTimer } from '../../../common/observableTimer';
+import {
+    getExperimentRootDir,
+} from '../../../common/utils';
+import { GPUSummary, ScheduleResultType } from '../../common/gpuData';
+import { TrialConfig } from '../../common/trialConfig';
+import { TrialConfigMetadataKey } from '../../common/trialConfigMetadataKey';
+import { execMkdir, validateCodeDir } from '../../common/util';
+import { GPUScheduler } from '../../remote_machine/gpuScheduler';
+import {
+    ExecutorManager, RemoteMachineMeta,
+    RemoteMachineScheduleInfo, RemoteMachineScheduleResult
+} from '../../remote_machine/remoteMachineData';
+import { ShellExecutor } from 'training_service/remote_machine/shellExecutor';
+import { RemoteMachineEnvironmentInformation, RemoteMachineMetaDetail } from '../remote/remoteConfig';
+
+
+@component.Singleton
+export class RemoteEnvironmentService extends EnvironmentService {
+
+    private readonly initExecutorId = "initConnection";
+    private readonly machineExecutorManagerMap: Map<RemoteMachineMeta, ExecutorManager>;
+    private readonly machineCopyExpCodeDirPromiseMap: Map<RemoteMachineMeta, Promise<void>>;
+    private readonly environmentExecutorManagerMap: Map<string, ExecutorManager>;
+    private readonly environmentJobsMap: Map<string, RemoteMachineEnvironmentInformation>;
+    private readonly expRootDir: string;
+    private trialConfig: TrialConfig | undefined;
+    private gpuScheduler?: GPUScheduler;
+    private readonly jobQueue: string[];
+    private readonly timer: ObservableTimer;
+    private stopping: boolean = false;
+    private readonly metricsEmitter: EventEmitter;
+    private readonly log: Logger;
+    private sshConnectionPromises: any[];
+    private experimentRootDir: string;
+    private experimentId: string;
+
+    constructor(@component.Inject timer: ObservableTimer) {
+        super();
+        this.experimentId = getExperimentId();
+        this.metricsEmitter = new EventEmitter();
+        this.environmentJobsMap = new Map<string, RemoteMachineEnvironmentInformation>();
+        this.environmentExecutorManagerMap = new Map<string, ExecutorManager>();
```

Comment:
Do you mean that we do not need to schedule environment? We doe not maintain machine in OpenPAI and AML mode because these platform have the schedule mechanism originally, while NNI need to maintain these machine in remote mode.

Reply:
If trial dispatcher asks an environment, the remote env service should initialize the environment, and return it back to trial dispatcher. scheduling in remote, is to return remote machine from the machine list one by one.